### ISSUE TITLE
OT is badly applied

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Starcounter#3769]
1. Let the initial state be `{val:0}`
1. Change it on server to `1` and send an update to client
1. Before the update reaches the client, change the state on client to `2` and send it to server
1. Let the update from server reach the client
1. Now both sides changed the value independently - OT should be applied and server version (`1`) should "win" and prevail on both sides
1. What happens is client accepts the change from server (correctly), but server also accepts the change from client (incorrectly), so they have different states (client has `1`, server has `2`), regardless of the fact that their versions match

I don't have a nice reproduction code for this, I reproduced it manually. I can help in creating reproduction code.

**Comment by miyconst** [comment]
Good catch @joozek78! I think this is something @tomalec and @chrhol should take a look at. Not related to `Typed JSON` though.

**Comment by tomalec** [comment]
ehh.. once again I miss some time to finally write a paper on JSON Patch OT

Even though Puppet protocol is symmetric, Transformations in our case are not. As we decided that "server-side JSON is the only source of truth" server-side operations should always be effectively included ("win"). See ET and IT at https://en.wikipedia.org/wiki/Operational_transformation#OT_functions.

If server-side transformations are works otherwise, it's a bug.

**Comment by tomalec** [comment]
BTW, at 5, 6 none of the sides "accepts" or "rejects" the change. The are both transforming the changes no matter what.

**Comment by tomalec** [comment]
You can use my 2yrs old tool http://tomalec.github.io/PuppetJs-operational-transformation/visualization.html to create patches (it's obviously not yet finished, and works with really old version of everything).

I'll translate @joozek78's example to the realt JSON Patches:

1. Browser (Alice) changes `smth` to `1` and sends it:

 ```js
A1 = [
     {"op":"test","path":"/_ClientVersion$","value":0},
     {"op":"replace","path":"/_ClientVersion$","value":1},
     {"op":"test","path":"/_ServerVersion","value":"0"},
     {"op":"replace","path":"/Smth","value":1}
]
```
2. Server (Bob) changes `smth` to `2` and sends it:

 ```js
B1 = [
     {"op":"test","path":"/_ServerVersion","value":0},
     {"op":"replace","path":"/_ServerVersion","value":1},
     {"op":"test","path":"/_ClientVersion$","value":"0"},
     {"op":"replace","path":"/Smth","value":2}
]
```
3. Browser (Alice) receives change from Bob (`B1`), and transforms it against hers (`A1`), to effectively include Bob's change (Bob is the only source of truth)

 ```js
B1′ = B1 * A1 = [
     {"op":"test","path":"/_ServerVersion","value":0},
     {"op":"replace","path":"/_ServerVersion","value":1},
     {"op":"test","path":"/_ClientVersion$","value":"1"}, // <-
     {"op":"replace","path":"/Smth","value":2} // <-
]
```

4. Server (Bob) receives change from Alice (`A1`), and transforms it against his (`B1`), to effectively include Bob's change (Bob is the only source of truth)

 ```js
A1′ = A1 * B1 = [
     {"op":"test","path":"/_ClientVersion$","value":0},
     {"op":"replace","path":"/_ClientVersion$","value":1},
     {"op":"test","path":"/_ServerVersion","value":"1"}, // <-
     {"op":"replace","path":"/Smth","value":2} // <-
]
```

Thanks to such design Server/Bob can theoretically use both `A1` and `A1′`. `A1` to just prompt a message, `A1′` as a representation of eventually-consistent state.

From what @joozek78 said, step 4 is buggy. as it does `B1*A1` instead of `A1*B1`.

**Comment by dan31** [comment]
CC @chrhol

**Comment by chrhol** [comment]
Thanks for the ping @dan31. Totally missed this issue. 

It's clearly a bug on the server. I think I missed this part, and the server simply applies the change without consideration of the version.
